proposals/20250923-plugin-system-event-schema-versioning.md

```diff
@@ -0,0 +1,169 @@
+
+**Determine if schema version check is needed:**
+- If the plugin provides parsing capabilities (`CAP_PARSING`) and consumes `syscall` events
+    - Check if `get_parse_event_types` is not defined or provides an empty array and `m_parse_event_sources` contains `"syscall"`
+    - Check if `get_parse_event_types` provides an array containing at least a driver event
+    - Do not check otherwise, because only plugin events are consumed
+- If the plugin provides field extraction capabilities (`CAP_EXTRACTION`) and consumes `syscall` events
+    - Check if `get_extract_event_types` is not defined or provides an empty array and `m_extract_event_sources` contains `"syscall"`
+    - Check if `get_extract_event_types` provides an array containing at least a driver event
+    - Do not check otherwise, because only plugin events are consumed
+- If the plugin only provides event sourcing capabilities (`CAP_SOURCING`) or async capabilities (`CAP_ASYNC`), no check is needed
+
+In this case, the event schema version check must be performed.
+
+### 3. Schema Version Compatibility Check
+
+In case the check is required, the following logic is applied:
+
+1. **Default schema version:**
+   - If the plugin does not implement the `get_required_schema_version` function,
+     it is assumed that it is compatible with schema version `3.0.0`
+
+2. **Version string validation:**
+   - If the plugin implements the `get_required_schema_version` function, the
+     returned version string is parsed and validated
+   - The version string must follow the semver format: `"<major>.<minor>.<patch>"`
+   - If the version string is malformed or cannot be parsed, the plugin load fails with an error
+
+3. **Version compatibility check:**
+   - The driver event schema version in use in the current Falco libs version is
+     compared with the required schema version declared by the plugin
+   - **Major version incompatibility:** If the major versions differ, the plugin load fails
+   - **Minor version incompatibility:** If the major versions match but the plugin requires a higher minor version than available, the plugin load fails
+   - **Patch version:** Patch version differences are allowed (backward compatible)
+
+4. **Error handling:**
+   - **Malformed version string:** `"plugin provided an invalid required schema version: '<version>'"`
+   - **Major version mismatch:** `"plugin required schema version '<version>' not compatible with the driver schema version '<version>': major versions disagree"`
+   - **Minor version mismatch:** `"plugin required schema version '<version>' not compatible with the driver schema version '<version>': driver schema minor version is less than the requested one"`
+
+### 4. Implementation Details
+
+The implementation will require the following changes:
+
+1. **Plugin API Extension:**
+   - Add `get_required_schema_version` function pointer to the `plugin_api` struct
+   - Place it after `get_required_api_version` for logical grouping
```

Comment:
This breaks binary compatibility with static plugins :( We can work around it but could we put it at the end of the struct instead?

Reply:
Actually, after giving a second thought I'm not sure about what's matter the most between compatibility of static plugins and API header file consistency and readability. Anyway I updated the proposal and we can decide at implementation time.